plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPropertyPage.java

```diff
@@ -56,81 +105,161 @@ protected Control createContents(Composite parent) {
+    setupAutoPromoteDataBinding(bindingContext);
+    setupBucketDataBinding(bindingContext);
+
+    PreferencePageSupport.create(this, bindingContext);
+    observables = new ObservablesManager();
+    observables.addObservablesFromContext(bindingContext, true, true);
+  }
+
+  private void setupProjectIdDataBinding(DataBindingContext context) {
+    ISWTObservableValue promptObservable = WidgetProperties.selection().observe(promptForProjectIdButton);
+    ISWTObservableValue projectIdObservable = WidgetProperties.text(SWT.Modify).observe(projectId);
+
+    context.bindValue(promptObservable, model.promptForProjectId);
+
+    context.addValidationStatusProvider(new ProjectIdMultiValidator(promptObservable, projectIdObservable));
+  }
+
+  private void setupProjectVersionDataBinding(DataBindingContext context) {
+    final ISWTObservableValue overrideObservable = WidgetProperties.selection().observe(overrideDefaultVersionButton);
+    final ISWTObservableValue versionObservable = WidgetProperties.text(SWT.Modify).observe(version);
+
+    context.bindValue(overrideObservable, model.overrideDefaultVersioning);
+    context.bindValue(versionObservable, model.version);
+    context.bindValue(WidgetProperties.enabled().observe(versionLabel), model.overrideDefaultVersioning);
+    context.bindValue(WidgetProperties.enabled().observe(version), model.overrideDefaultVersioning);
+
+    context.addValidationStatusProvider(new OverrideValidator(overrideObservable,
+                                                              versionObservable,
+                                                              new ProjectVersionValidator()));
+  }
+
+  private void setupAutoPromoteDataBinding(DataBindingContext context) {
+    context.bindValue(WidgetProperties.selection().observe(autoPromoteButton), model.autoPromote);
+  }
+
+  private void setupBucketDataBinding(DataBindingContext context) {
+    final ISWTObservableValue overrideObservable = WidgetProperties.selection().observe(overrideDefaultBucketButton);
+    final ISWTObservableValue bucketObservable = WidgetProperties.text(SWT.Modify).observe(bucket);
+
+    context.bindValue(overrideObservable, model.overrideDefaultBucket);
+    context.bindValue(bucketObservable, model.bucket);
+    context.bindValue(WidgetProperties.enabled().observe(bucketLabel), model.overrideDefaultBucket);
+    context.bindValue(WidgetProperties.enabled().observe(bucket), model.overrideDefaultBucket);
+
+    context.addValidationStatusProvider(new OverrideValidator(overrideObservable,
+                                                              bucketObservable,
+                                                              new BucketNameValidator()));
+  }
+
+  @Override
+  public boolean performOk() {
+    if (isValid()) {
+      try {
+        savePreferences();
+        return true;
+      } catch (IOException exception) {
+        logger.log(Level.SEVERE, "Could not save deploy preferences", exception);
```

Comment:
this failure, unlikely as it is, probably calls for more than a log. We should alert the user the save failed, and maybe not close the dialog.